ships/0021-local-source-upload.md

```diff
@@ -0,0 +1,348 @@
+desired pod.
+
+Analogous to `kubectl cp`, we will be using `tar` to stream data from the client, while
+<code>[kubectl exec][KubectlExec]</code> on the pod keeps a <code>tar</code> process waiting for
+data on <code>STDIN</code>.
+
+#### Shipwright Build Controller
+
+The Shipwright Build Controller works based on [Tekton Workspaces][TektonWorkspaces], and therefore,
+it already defines the /workspace/source mount point for all containers in the build-pod. Tekton also
+adds it's own helper to ["git clone" the repository][ControllerGitClone] in place.
+
+With that said, we can simply stop Shipwright Build-Controller from issuing the "`git clone`", and
+instead, add a [simple helper application][ControllerWaiter] that's able to hold the build-pod
+execution while the data is transferred.
+
+To accomplish this the Build-Controller needs to be able to detect it should replace the "`git
+clone`" procedure by waiting for the upload helper instead. The changes done during research and
+development are [here][ControllerWaiterUsage].
+
+##### BuildRun Changes
+
+The proposed API changes are the following:
+
+```yml
+---
+apiVersion: shipwright.io/v1alpha1
+kind: BuildRun
+metadata: {}
+spec:
+  sources:
+    - name: src
+      type: LocalCopy
+```
+
+The Build Controller will inspect for a type `LocalCopy` on the `"sources"` slice, and therefore
+instead of adding the git clone step, it will instead wait for the user to upload.
+
+The `sources` slice is part of [Remote Artifacts (EP)][RemoteArtifactsEP] and has been worked on, so
+more effort may be needed to create a good API fit.
+
+In order to let each EP implementation stand-alone, an alternative mechanism should take place, by
+creating a simple annotation in the `BuildRun` object the Build Controller can be notified about the
+type of build expected. As per:
+
+```yml
+---
+apiVersion: shipwright.io/v1alpha1
+kind: BuildRun
+metadata:
+  annotations:
+    build.shipwright.io/upload.wait: true
+    build.shipwright.io/upload.wait.timeout: 120
+spec: {}
+```
+
+Therefore when `build.shipwright.io/wait.upload` is present the waiting will take place, and by
+expressing `build.shipwright.io/upload.wait.timeout` the actual timeout is configured.
+
+#### Waiting Mechanism
+
+The container-image built by Shipwright is created by a series of steps running sequentially, those
+steps are executed in the build-pod, running each container and inspecting for return-code and
+errors.
+
+The task at hand, waiting for the users' data transfer, will follow the same basic structure, which
+means a simple application meant to wait for a specific condition, and exit gracefully afterwards.
+Here we may keep checking if a given "lock" file exists, when removed the application exists. It also
+must exit with an error on timeout.
+
+Additionally, we can employ this application to make verifications on the streamed data, like for
+instance checksum, ownership (UID/GID) and more.
```

Comment:
I was also wondering about ownership of the files after upload. I guess they will be owned by the same user as locally. Another reasonable behavior might be to make them owned by the user who runs this container. But I have no arguments for the one or other now.

Reply:
another candidate IMO for an annotation to allow choice here until we get some user feedback and possible consensus, or lack there of, for the preference

where then if a consensus forms, we fee more confident on either providing "the one way to do it", or if a couple of finite choices prevail, we have the input for a first class api field